### DISCUSSION TITLE
SoftwareReport creation

### DISCUSSION THREAD

**Discussion by chrisdecker1201** [actions/virtual-environments#6587] [Ideas]
Hello,

I'm starting to use this repository as a template for my needs. As I'm only need a subset of supported languages and want to speed up the image creation by reducing unneeded components.

## I give a short example for what my issue is

When I want, to remove for example the `nvm` and `node` component,  I can simply remove these two lines in the `ubuntu2204.pkr.hcl` file

~~~diff
-"${path.root}/scripts/installers/nvm.sh",
-"${path.root}/scripts/installers/nodejs.sh",
~~~

But additionally, I've also to remove the following in the `SoftwareReport.Generator.ps1` script

~~~diff
-(Get-NodeVersion),
-(Get-NvmVersion),
~~~

## My idea

I think that for the most (maybe all) components, you've a test which check if the component is installed correctly. In the node example

~~~powershell
"$NodeCommand --version" | Should -ReturnZeroExitCode
~~~

In the `SoftwareReport.Common.psm1` script, you get the version with the following function

~~~powershell
function Get-NodeVersion {
    $nodeVersion = $(node --version).Substring(1)
    return "Node $nodeVersion"
}
~~~

Wouldn't it be nice to create some kind of report file during/after the installation & test process, which contains the information that are already checked?

I think about something like the following workflow.

1. Create a report script, where you can give the context of where to place this in the markdown
2. Call this script after the invoke test script. E.g.:
~~~bash
report-component --topic "Installed Software" --subtopic "Language and Runtime" --component "Node" --version "16.18.0" --cached-tools {"title": "Node.js", "tools":["14.20.1", "16.18.0", "18.12.0"]}
~~~
3. Add all this data to a `report.json` file
~~~
{
    "Installed Software": {
        "Language and Runtime": {
            "Node": "16.18.0",
        }
        "Cached Tool": {
             "Node.js": [
                  "14.20.1",
                  "16.18.0",
                  "18.12.0"
             ]
        }
}
~~~
5. Combine all `report.json` to one file.
4. Create the Markdown from this JSON-Object

This will make it easier to add/remove components because you will have only the installation script with install, test & report.
Additionally, this could also decrease build time a little bit, if you do not call every component twice for version information.

What do you think?

**Comment by mikhailkoliada** [comment]
Hello! I do not understand how it would ease the maintenance (or am I missing the idea?) We currently generate markdown from powershell directly, creating it from json is also possible, but has never been really considered.

**Comment by chrisdecker1201** [comment]
Maybe that's an edge case or maybe I do not understand everything correct :).

What I'm understanding is that you have the following workflow:

1. Install everything and test this
2. Create Report for everything

My idea was the following workflow:

1. Install a component, test this component and create a report for this component.
2. Combine all reports to a single component.

At the moment, you have to add/remove a component at two places. Add/Remove install script add/remove report creation.

Probably that's not an issue for you, but I want to create an image depending on this repository without every component to speed up the image creation, and therefore I've to delete the installation as well as the report creation.
